### PR TITLE
Ap/parameterization

### PR DESCRIPTION
**Pull request by austinhpatton**
This PR includes multiple changes that convert numerous previously hard-coded software parameters into actual nextflow parameters. These are now specified either in the parameter file (json), or in the modules.config file. Default parameters are now specified in base.config. 

This PR corresponds to the following issue, which outlines the problem, and for each fix indicates where/how parameterization was conducted: https://github.com/Arcadia-Science/phylorthology/issues/27. 

Changes to iqtree were most extensive - previously the user was forced to use a complex model of amino acid substitution, which involved running iqtree once, inferring a guide tree, and running a second time using that tree. I now include two parameters for iqtree: tree_model, and tree_model_pmsf. The first has a default (iqtree will automatically determine this) whereas the second does not. The second stage (for pmsf) will only run if this is specified in the parameterfile.

Additionally, I include a check in the "orthofinder prep" module that will only write out the sequence files (potentially >10-20 thousands of them per value of MCL inflation) if we are not at the MCL testing stage and are running the analysis in full. This will save tons of storage space and make cleanup more straightforward.

### CODE REVIEW

File: conf/base.config
Comment (mertcelebi) [inline conf/base.config:12]: These should be in the `nextflow.config` file not here. That's where all the inputs are defined and we should keep that consistent.

File: workflows/phylorthology.nf
Comment (mertcelebi) [inline workflows/phylorthology.nf:279]: it's been a while since i looked at these, are you sure about this?

File: conf/base.config
Comment (mertcelebi) [inline conf/base.config:17]: should the default be something else?

File: modules/local/nf-core-modified/iqtree.nf
Comment (mertcelebi) [inline modules/local/nf-core-modified/iqtree.nf:47]: We want this to be additive to the `iqtree` run above right? Not an if/else situation?

File: modules/local/nf-core-modified/iqtree.nf
Comment (mertcelebi) [inline modules/local/nf-core-modified/iqtree.nf:49]: hate this lol

**Review by mertcelebi — CHANGES_REQUESTED** [review]
Looks good to me, very close to finished. Left a couple questions, please also update the `nextflow_schema.json` file with the inputs.

File: workflows/phylorthology.nf
Comment (austinhpatton) [reply workflows/phylorthology.nf:122]: Not sure. Must be some leftover from attempts to parameterize this. Can probably just be deleted, since it runs as it should.

File: workflows/phylorthology.nf
Comment (austinhpatton) [reply workflows/phylorthology.nf:126]: ....cause I felt like I should? I don't actually have strong feelings about this - seems like you tend to be inclined to just directly access the outputs from each module as needed rather than setting these as channels. Curious if there's an underlying reason for this? Could there be cases where outputting things like these into channels can introduce problems of their own?

File: conf/base.config
Comment (austinhpatton) [reply conf/base.config:12]: Ahhh, okay - I wasn't sure where best to put these, but it worked so I stuck with it. Can move them easily though.

File: conf/base.config
Comment (austinhpatton) [reply conf/base.config:17]: What do you mean? 

As is, I've actually tried to specify some sensible defaults - I don't actually intend to use these myself, but they are are least less sensitive to specific details of any one dataset.

File: modules/local/nf-core-modified/iqtree.nf
Comment (austinhpatton) [reply modules/local/nf-core-modified/iqtree.nf:49]: Me too - we should be able to use the normal "auto" method for detecting number of threads, but it fails for whatever reason when running the PMSF approximation. That bug is internal to IQ-Tree. So, this is just a workaround to still use a sensible number of threads from the previous step. 

I might end up looking into whether the tree estimation can actually slow down from overspecification of threads, or if it's just a question of diminishing returns. If the latter, I could imagine just specifying a fixed number of threads (as a parameter, with the auto method as default) if we're running on tower for instance.

File: modules/local/nf-core-modified/iqtree.nf
Comment (austinhpatton) [reply modules/local/nf-core-modified/iqtree.nf:47]: It's admittedly a bit weird. We do want it to be an if-else.

Basically, if you only specify one model (the "tree_model" parameter), and "pmsf_model" is not specified, and equals "none" by default, then you just run iqtree once. If you've specified an actual second model "pmsf_model", then you run what's in this second step. 

I'm just using the if/else statement as a way to safeguard against needlessly inferring two trees, unless that's explicitly the intent of the user.

File: workflows/phylorthology.nf
Comment (austinhpatton) [reply workflows/phylorthology.nf:279]: Yup!

**Comment by austinhpatton** [comment]
> Looks good to me, very close to finished. Left a couple questions, please also update the `nextflow_schema.json` file with the inputs.

Great! Working on making those changes and will push them soon - hopefully can get this merged today. Keep forgetting about that schema file...

Also, I ended up needing to make one additional small change - we ended up tailoring the workflow to work using data where all species are from uniprot (previously worked because our test dataset was like this). As a result, if any species _wasn't_ from uniprot, the annotation module would fail, because the "accessions" output (among others) wouldn't be written. 

So, I modified the input_check subworkflow slightly so that it outputs a channel of just the species for which we can annotate proteins - that channel is used as input to the annotation module.

**Comment by austinhpatton** [comment]
> Great fix with adding parameter arguments instead of hardcoded options. The one thing I think that can be addressed later is usually the amount of information that you've put into the README about the parameters is goes into the nextflow_schema.json file where you can put information about different parameters and you can edit that file interactively with the nf-core tools. Additionally if you have time (and you want to) the local modules you have that aren't available through nf-core would be cool if you could make PRs to contribute those modules there, but not necessary.

Thanks! And that's good to know - hadn't realized that was the normal/correct way to do this, nor that you could do it interactively with nf-core tools. 

But yeah, as for the modifications to nf-core, it's something we've discussed, and may end up doing later on for some modules in particular, but some of the changes are kind of customized/oriented towards my specific use case, and so might not be useful in a broader sense. Case by case basis though!

File: conf/base.config
Comment (mertcelebi) [reply conf/base.config:17]: What I mean is: if someone ran this with the default value of "TEST", will the pipeline fail or run in a semi-intelligent way? If not, we should either have a sensible default or not have a default and throw an error if not specified.

File: conf/base.config
Comment (mertcelebi) [reply conf/base.config:12]: Don't forget to move these.

**Review by mertcelebi — COMMENTED** [review]
All the new changes look good. But don't forget to update the `nextflow.schema` and the move the base config changes. Once those are in, ping me for a final approval.

File: conf/base.config
Comment (austinhpatton) [reply conf/base.config]: Yeah, this is actually the most foolproof. Basically it makes it so that it's agnostic to whether input sequences are amino acids or nucleotides. I suppose since I haven't tested the workflow with nucleotides, I could just set a sensible model for AA sequences?

File: conf/base.config
Comment (austinhpatton) [reply conf/base.config]: Done.

File: conf/base.config
Comment (mertcelebi) [inline conf/base.config:18]: good stuff

File: modules/local/annotate_uniprot.nf
Comment (mertcelebi) [inline modules/local/annotate_uniprot.nf:9]: love to see it

File: modules/local/nf-core-modified/mafft.nf
Comment (mertcelebi) [inline modules/local/nf-core-modified/mafft.nf:30]: hmm we sure about this?

File: nextflow.config
Comment (mertcelebi) [inline nextflow.config:32]: Asking this one last time so I am sure:
> What I mean is: if someone ran this with the default value of "TEST", will the pipeline fail or run in a semi-intelligent way? If not, we should either have a sensible default or not have a default and throw an error if not specified.

**Review by mertcelebi — APPROVED** [review]
Approving with the caveat that the answers to my two questions are not terrible :) I'll keep the nextflow_schema as is for now and fix it myself later. So no need to hold up the PR. The rest looks good.

File: conf/base.config
Comment (austinhpatton) [reply conf/base.config:18]: Yup, much better this way

File: modules/local/annotate_uniprot.nf
Comment (austinhpatton) [reply modules/local/annotate_uniprot.nf:9]: All part of what's coming next with the README/publishdir changes!

File: modules/local/nf-core-modified/mafft.nf
Comment (austinhpatton) [reply modules/local/nf-core-modified/mafft.nf:30]: I think it's safter than what I was doing. As is, I was specifying a fixed number of threads, but mafft should be able to autoscale for memory at the memory intensive stages in a sensible way with this flag.

File: nextflow.config
Comment (austinhpatton) [reply nextflow.config:32]: Thought I posted a response that seems to be missing for whatever reason. 

This makes it generally useable, but perhaps slower if something more pointed is specified. Basically, this means it can be run with either nucleotide or amino acid sequences using defaults. But, I haven't tested the whole workflow with nucleotides, so maybe I should just specify something specific to amino acids that's sensible, and I can point to this in the readme, eventually building out nucleotide usage in later versions?

....I'm going to do that - I've convinced myself (the second time writing this response, lol).

**Pull request merged**
